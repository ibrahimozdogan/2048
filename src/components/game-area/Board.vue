<template>
    <div class="block-table">
        <BlockRow
            v-for="(blocks, rowIndex) in blockRows"
            :row-index="rowIndex"
            :blocks="blocks"
            :key="rowIndex"
        />
        <div v-if="this.blockRows.length === 0" class="block-table-text">
            {{translations.blockTableText}}
        </div>
    </div>
</template>

<script>
    import { mapActions, mapGetters } from 'vuex';
    import CommonEnums from '../../enums/CommonEnums';
    import MatrixService from '../../utils/MatrixService';
    import StorageService from '../../utils/StorageService';
    import Swipe from '../../utils/Swipe';
    import BlockRow from '../block-row/Index';

    const storageService = new StorageService();

    export default {
        components: {
            BlockRow
        },

        mounted () {
            this.listenKeyUpEvents();
            new Swipe().listen(this.shift);

            this.setProperty({ key: 'bestScore', value: storageService.get('best-score') });
        },

        data () {
            return {
                translations: {
                    blockTableText: 'Click `New Game` button to start game'
                }
            }
        },

        computed: {
            ...mapGetters({
                blockRows: 'getBlockRows',
                currentScore: 'getCurrentScore',
            })
        },

        methods: {
            ...mapActions([
                'setProperty',
                'incrementScore'
            ]),

            /**
             * @param {string} direction
             */
            shift (direction) {
                const matrixService = new MatrixService({
                    matrix: this.blockRows,
                    newElementValue: CommonEnums.NEW_ELEMENT_VALUE
                });
                const { matrix, status, accumulatedValue } = matrixService.merge(direction);

                if (status) {
                    matrixService.addRandomValue(1);
                }

                this.setProperty({ key: 'blockRows', value: matrix });
                this.incrementScore(accumulatedValue);

                if (this.currentScore > (storageService.get('best-score') || 0)) {
                    storageService.set({ key: 'best-score', value: this.currentScore });
                    this.setProperty({ key: 'bestScore', value: this.currentScore });
                }
            },

            listenKeyUpEvents () {
                document.addEventListener('keyup', event => {
                    const key = event.key;
                    const direction = key.replace('Arrow', '').toLowerCase();

                    if (CommonEnums.ALLOWED_BUTTONS.includes(key)) {
                        this.shift(direction)
                    }
                });
            }
        }
    };
</script>

<style lang="scss" scoped>
    $dark: #253342;
    $background-color: #BEAD9E;
    $text-color: #FEFEFE;
    $table-size: 296px;
    $display: flex;
    $center: center;


    .block-table {
        border-radius: 3px;
        background: $background-color;
        min-width: $table-size;
        min-height: $table-size;
        position: relative;
        padding: 5px;

        .block-table-text {
            color: $text-color;
            position: absolute;
            left: 50%;
            top: 50%;
            transform: translate(-50%, -50%);
        }
    }
</style>
