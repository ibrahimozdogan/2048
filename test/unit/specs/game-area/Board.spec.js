import { shallowMount, createLocalVue } from '@vue/test-utils'
import Board from '@/components/game-area/Board'
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('@components/game-area/Board.vue', () => {
    let gameStore;
    let store;
    let wrapper;
    let data;
    let currentScore = 10;
    let bestScore = 20;
    let blockRows = [[0, 2, 2048, 4], [0, 2, 0, 4], [0, 2, 0, 4], [0, 2, 0, 4]];

    beforeEach(() => {
        gameStore = {
            actions: {
                setProperty: jest.fn(),
            },
            getters: {
                getCurrentScore: () => currentScore,
                getBestScore: () => bestScore,
                getBlockRows: () => blockRows,
            },
            state: { currentScore, bestScore, blockRows },
        };
        store = new Vuex.Store({ modules: { gameStore } });
        data = {
            translations: {
                blockTableText: 'Click `New Game` button to start game'
            }
        };
        wrapper = shallowMount(Board, {
            store,
            localVue,
            propsData: { data }
        });
    });

    describe('render', () => {
        it('renders correctly', () => {
            expect(wrapper.element).toMatchSnapshot();
        });
    });
});
