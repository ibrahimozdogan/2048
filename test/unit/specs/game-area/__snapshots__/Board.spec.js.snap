// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/game-area/Board.vue render renders correctly 1`] = `
<div
  class="block-table"
  data="[object Object]"
>
  <blockrow-stub
    blocks="0,2,2048,4"
    rowindex="0"
  />
  <blockrow-stub
    blocks="0,2,0,4"
    rowindex="1"
  />
  <blockrow-stub
    blocks="0,2,0,4"
    rowindex="2"
  />
  <blockrow-stub
    blocks="0,2,0,4"
    rowindex="3"
  />
   
  <!---->
</div>
`;
